### PR TITLE
feat: 3D/spatial audio

### PR DESCRIPTION
**Pull request by jullianpalma**
`AudioService` can now create and play sources in 3D space.

made 2 new components for a `Listener` and a `SoundEmitter`, both of which will continuously update the location of whatever entity they're attached to to the `AudioService`.

also added a handful of sound effects in `resources/audio/` and renamed a couple.

### CODE REVIEW

File: game/src/engine/audio/AudioService.h [hunk 1]
```diff
@@ -3,81 +3,152 @@
...
      */
     void SetPitch(std::string file_name, float pitch_offset);
 
+    /**
+     *  set the world position of a source to play from.
+     *
+     *  @param entity_id the sources associated entity id.
+     *  @param position the position to set the source at.
+     */
+    void SetSourcePosition(std::int32_t entity_id, glm::vec3 position);
```
Comment (raduschirliu) [inline game/src/engine/audio/AudioService.h:134]: Missing a `u` here, Entity ID should be a `uint32_t`

File: game/src/game/components/audio/Listener.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include "engine/audio/AudioService.h"
+#include "engine/scene/Component.h"
+#include "engine/scene/OnUpdateEvent.h"
+#include "engine/scene/Transform.h"
+
+class Listener final : public Component, public IEventSubscriber<OnUpdateEvent>
```
Comment (raduschirliu) [inline game/src/game/components/audio/Listener.h:8]: Could we rename this to `AudioListener` instead? if you're just looking at the class name when it's added as a component it's a bit hard to tell if it's related to events or not. imo `Listener` could be mistaken as being the same as `EventSubscriber`,  i.e. `EventListener`)

File: game/src/engine/audio/AudioService.cpp [hunk 3]
```diff
@@ -22,75 +25,359 @@ const ALCchar* kDefaultDevice = nullptr;
...
 
+// the current music file to stream from
+AudioFile kMusicAudioFile;
+
+// just keeps track of how much of the file was played
+ALsizei kMusicPlayhead = 0;
+
+// number of buffers to use when streaming music files
+const std::size_t kStreamBufferAmount = 4;
+const ALsizei kStreamBufferSize = 65536;  // 32kb per buffer
```
Comment (raduschirliu) [inline game/src/engine/audio/AudioService.cpp:39]: These could be `static constexpr` to make them compile time constants only for this file instead of just general constant variables

File: game/src/engine/audio/AudioService.cpp [hunk 4]
```diff
@@ -22,75 +25,359 @@ const ALCchar* kDefaultDevice = nullptr;
...
-// note: since diagetic sound (3d sound) can only be in mono
-// we'll want all sfx to be mono too (music is fine in stereo)
 const std::string kSfxDirectory = "resources/audio/sfx/";
 const std::string kMusicDirectory = "resources/audio/music/";
 
+// the current music file to stream from
+AudioFile kMusicAudioFile;
+
+// just keeps track of how much of the file was played
+ALsizei kMusicPlayhead = 0;
```
Comment (raduschirliu) [inline game/src/engine/audio/AudioService.cpp:35]: Could these be member vars of AudioService instead of being scoped to the file?

File: game/src/engine/audio/AudioService.cpp [hunk 5]
```diff
@@ -22,75 +25,359 @@ const ALCchar* kDefaultDevice = nullptr;
 ALCint kContextAttributes = 0;
 
 // audio directories
-// note: since diagetic sound (3d sound) can only be in mono
-// we'll want all sfx to be mono too (music is fine in stereo)
 const std::string kSfxDirectory = "resources/audio/sfx/";
 const std::string kMusicDirectory = "resources/audio/music/";
```
Comment (raduschirliu) [inline game/src/engine/audio/AudioService.cpp:29]: Usually it's a good idea to add `static` to variables that are meant to only exist in one source file/compilation unit. reason being, just by looking at these vars here, there's no way to tell if they're only meant for this file or if there's a header file somewhere defining like:
```c++
extern ALCint kContextAttributes
```
If a header _did_ exist like this, these vars would then refer to the thing in the header.

tl;dr: being explicit usually makes it easier to determine what stuff means when looking at it later :P

File: game/src/engine/audio/AudioFile.h [hunk 6]
```diff
@@ -7,8 +7,15 @@
...
     int number_of_channels_;
-    int samples_per_channel;
-    int sample_rate_;
+    ALuint samples_per_channel;
+    ALsizei sample_rate_;
     ALenum format_;
-    std::unique_ptr<short> data_;
+    std::unique_ptr<ALshort> data_;
+
+    ALsizei get_size_bytes()
```
Comment (raduschirliu) [inline game/src/engine/audio/AudioFile.h:15]: iirc the actual Google Style Guide does say to do getters/setters in snake_case instead of UpperCamel, but I think all other getters/setters in our code are currently UpperCamel, so maybe rename to `GetSizeBytes()` for consistency?

**Review by raduschirliu — CHANGES_REQUESTED** [review]
Looks great, just some minor comments! And damn you weren't kidding when you said this ended up being a lot of work :0

File: game/src/engine/audio/AudioService.cpp (hunk 3 above)
Comment (jullianpalma) [reply game/src/engine/audio/AudioService.cpp:39]: ooh, got it! wasn't actually aware of `constexpr` until just right now lol

File: game/src/engine/audio/AudioFile.h (hunk 6 above)
Comment (jullianpalma) [reply game/src/engine/audio/AudioFile.h:15]: whoops! force of habit, usually i catch these before pushing !

File: game/src/game/components/audio/Listener.h (hunk 2 above)
Comment (jullianpalma) [reply game/src/game/components/audio/Listener.h:8]: yeah, i was thinking this too

File: game/src/engine/audio/AudioService.cpp (hunk 3 above)
Comment (raduschirliu) [reply game/src/engine/audio/AudioService.cpp]: Haha yep, I also didn't know it existed for the longest time :P some modern C++ stuff has a tendency to be a bit more hidden sadly

Fun fact, there's also an `if constexr (...) {...}` which can sometimes replace an `#ifdef`.

**Review by raduschirliu — APPROVED** [review]
